File: soac.go
package soac

import (
	"fmt"
)

type Color byte
type Attribute byte
type BackGround byte

type AttrFactor interface {
	Get() int
}

const (
	Black Color = iota + 30
	Red
	Green
	Yellow
	Blue
	Magenda
	Cyan
	White
)

func (c Color) Get() int {
	return int(c)
}

const (
	Reset Attribute = iota
	Bold
	Faint
	Italic
	Underline
	Blink1
	Blink2
	Reverse
	Concealed
	Crossedout
)

func (a Attribute) Get() int {
	return int(a)
}

const (
	BgBlack BackGround = iota + 40
	BgRed
	BgGreen
	BgYellow
	BgBlue
	BgMagenda
	BgCyan
	BgWhite
)

func (b BackGround) Get() int {
	return int(b)
}

type Changer struct {
	fg   Color
	attr Attribute
	bg   BackGround
}

func NewChanger() *Changer {
	return &Changer{39, 0, 49}
}

func (self *Changer) Apply(val interface{}) string {
	return fmt.Sprintf("\x1b[%d;%d;%dm%v\x1b[0m", self.attr, self.fg, self.bg, val)
}

func (self *Changer) Set(factors ...AttrFactor) {
	for _, factor := range factors {
		i := factor.Get()
		if 0 <= i && i <= 9 {
			self.attr = Attribute(i)
		} else if 30 <= i && i <= 37 {
			self.fg = Color(i)
		} else if 40 <= i && i <= 47 {
			self.bg = BackGround(i)
		} else {
			// return error or emit warnning
		}
	}
}

func (self *Changer) Reset() {
	self.fg = 39
	self.attr = 0
	self.bg = 49
}

func (self *Changer) Black() *Changer {
	self.fg = Black
	return self
}

func (self *Changer) Red() *Changer {
	self.fg = Red
	return self
}

func (self *Changer) Green() *Changer {
	self.fg = Green
	return self
}

func (self *Changer) Yellow() *Changer {
	self.fg = Yellow
	return self
}

func (self *Changer) Blue() *Changer {
	self.fg = Blue
	return self
}

func (self *Changer) Magenda() *Changer {
	self.fg = Magenda
	return self
}

func (self *Changer) Cyan() *Changer {
	self.fg = Cyan
	return self
}

func (self *Changer) White() *Changer {
	self.fg = White
	return self
}

func (self *Changer) Bold() *Changer {
	self.attr = Bold
	return self
}

func (self *Changer) Faint() *Changer {
	self.attr = Faint
	return self
}

func (self *Changer) Italic() *Changer {
	self.attr = Italic
	return self
}

func (self *Changer) Underline() *Changer {
	self.attr = Underline
	return self
}

func (self *Changer) Blink1() *Changer {
	self.attr = Blink1
	return self
}

func (self *Changer) Blink2() *Changer {
	self.attr = Blink2
	return self
}

func (self *Changer) Reverse() *Changer {
	self.attr = Reverse
	return self
}

func (self *Changer) Concealed() *Changer {
	self.attr = Concealed
	return self
}

func (self *Changer) CrossedOut() *Changer {
	self.attr = Crossedout
	return self
}

func (self *Changer) Bblack() *Changer {
	self.bg = BgBlack
	return self
}

func (self *Changer) Bred() *Changer {
	self.bg = BgRed
	return self
}

func (self *Changer) Bgreen() *Changer {
	self.bg = BgGreen
	return self
}

func (self *Changer) Byellow() *Changer {
	self.bg = BgYellow
	return self
}

func (self *Changer) Bblue() *Changer {
	self.bg = BgBlue
	return self
}

func (self *Changer) Bmagenda() *Changer {
	self.bg = BgMagenda
	return self
}

func (self *Changer) Bcyan() *Changer {
	self.bg = BgCyan
	return self
}

func (self *Changer) Bwhite() *Changer {
	self.bg = BgWhite
	return self
}

File: soac.py
class COLOR:
    BLACK   = 30
    RED     = 31
    GREEN   = 32
    YELLOW  = 33
    BLUE    = 34
    MAGENDA = 35
    CYAN    = 36
    WHITE   = 37

class ATTRIBUTE:
    RESET      = 0
    BOLD       = 1
    FAINT      = 2
    ITALIC     = 3
    UNDERLINE  = 4
    BLINK1     = 5
    BLINK2     = 6
    REVERSE    = 7
    CONCEALED  = 8
    CROSSEDOUT = 9

class BG:
    BLACK   = 40
    RED     = 41
    GREEN   = 42
    YELLOW  = 43
    BLUE    = 44
    MAGENDA = 45
    CYAN    = 46
    WHITE   = 47

class Changer:
    def __init__(self):
        self.color = 39
        self.attr  = 0
        self.bg    = 49

    def apply(self, val):
        return "\x1b[%d;%d;%dm%s\x1b[0m" % (self.attr, self.color, self.bg, val) 

    def black(self):
        self.color = COLOR.BLACK
        return self

    def red(self):
        self.color = COLOR.RED
        return self

    def green(self):
        self.color = COLOR.GREEN
        return self

    def yellow(self):
        self.color = COLOR.YELLOW
        return self

    def blue(self):
        self.color = COLOR.BLUE
        return self

    def magenda(self):
        self.color = COLOR.MAGENDA
        return self

    def cyan(self):
        self.color = COLOR.CYAN
        return self

    def white(self):
        self.color = COLOR.WHITE
        return self

    def reset(self):
        self.attr = ATTRIBUTE.RESET
        return self

    def bold(self):
        self.attr = ATTRIBUTE.BOLD
        return self

    def faint(self):
        self.attr = ATTRIBUTE.FAINT
        return self

    def italic(self):
        self.attr = ATTRIBUTE.ITALIC
        return self

    def underline(self):
        self.attr = ATTRIBUTE.UNDERLINE
        return self

    def blink1(self):
        self.attr = ATTRIBUTE.BLINK1
        return self

    def blink2(self):
        self.attr = ATTRIBUTE.BLINk2
        return self

    def reverse(self):
        self.attr = ATTRIBUTE.REVERSE
        return self

    def concealed(self):
        self.attr = ATTRIBUTE.CONCEALED
        return self

    def crossedout(self):
        self.attr = ATTRIBUTE.CROSSEDOUT
        return self

    def b_black(self):
        self.bg = BG.BLACK
        return self

    def b_red(self):
        self.bg = BG.RED
        return self

    def b_green(self):
        self.bg = BG.GREEN
        return self

    def b_yellow(self):
        self.bg = BG.YELLOW
        return self

    def b_blue(self):
        self.bg = BG.BLUE
        return self

    def b_magenda(self):
        self.bg = BG.MAGENDA
        return self

    def b_cyan(self):
        self.bg = BG.CYAN
        return self

    def b_white(self):
        self.bg = BG.WHITE        
        return self
